plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,434 @@
+#include "wayfire/core.hpp"
+
+#include "wayfire/seat.hpp"
+#include "wayfire/scene-input.hpp"
+#include "wayfire/scene-operations.hpp"
+#include "wayfire/unstable/wlr-surface-node.hpp"
+#include "wayfire/unstable/wlr-surface-controller.hpp"
+#include "wayfire/output.hpp"
+#include "wayfire/output-layout.hpp"
+#include "wayfire/signal-definitions.hpp"
+
+#include <wayfire/plugin.hpp>
+#include <wayfire/util.hpp>
+#include <wayfire/nonstd/wlroots-full.hpp>
+#include <wayfire/plugins/common/simple-text-node.hpp>
+#include <wayfire/util/log.hpp>
+
+enum lock_state
+{
+    LOCKING,
+    LOCKED,
+    UNLOCKED,
+    DESTROYED,
+    ZOMBIE,
+} state;
+
+class wf_session_lock_plugin : public wf::plugin_interface_t
+{
+    class wayfire_session_lock;
+
+    class lock_surface_node_t : public wf::scene::wlr_surface_node_t
```

Comment:
Note that all nodes are shared_ptrs - this means that in general, it is possible that a node outlives the session-lock plugin to begin with, it can also outlive the output it is on, etc.

So, you should try to add code so that the object can 'work' (at least not crash) even as the underlying surface, lock, output, etc. are destroyed. I have tried to add some suggestions on how to accomplish this.